File: frontend/src/data/writingTools.js
export const EDIT_STRENGTHS = [
  { id: 'light', label: 'Light edit', hint: 'Fix only clear grammar and word choice problems' },
  { id: 'standard', label: 'Standard edit', hint: 'Fix grammar, clarity, and naturalness' },
  { id: 'strong', label: 'Strong rewrite', hint: 'Improve flow, structure, and clarity' },
  { id: 'teacher', label: 'Teacher mode', hint: 'Correct, explain, and give examples' },
]

export const EDIT_STYLES = [
  { id: 'simple_american_english', label: 'Simple American English', hint: 'Clear, everyday American English for learners' },
  { id: 'academic_english', label: 'Academic English', hint: 'Formal tone, precision, and paragraph structure' },
  { id: 'toefl_writing', label: 'TOEFL Writing', hint: 'Organized, direct writing for TOEFL practice' },
  { id: 'email_professional', label: 'Email / Professional', hint: 'Polite, concise workplace or academic email' },
  { id: 'natural_conversation', label: 'Natural Conversation', hint: 'Friendly, spoken-style everyday English' },
]

export const EDIT_LANGUAGE_LEVELS = [
  { id: 'beginner', label: 'Beginner', hint: 'Short sentences, common words, simple grammar (A1–B1)' },
  { id: 'normal', label: 'Normal', hint: 'Natural sentences with moderate complexity (B1–B2)' },
  { id: 'professional', label: 'Professional', hint: 'Polished, precise vocabulary and smoother structure' },
]

const STYLE_ALIASES = {
  simple: 'simple_american_english',
  academic: 'academic_english',
  toefl: 'toefl_writing',
  email: 'email_professional',
  conversation: 'natural_conversation',
}

export function normalizeEditStyle(style) {
  return STYLE_ALIASES[style] || style || 'simple_american_english'
}

export function normalizeEditLanguageLevel(level) {
  const key = (level || 'normal').toLowerCase()
  if (['beginner', 'normal', 'professional'].includes(key)) return key
  return 'normal'
}

export function getEditOption(list, id) {
  return list.find((item) => item.id === id) || list[0]
}

export function describeEditSettings({ strength, style, languageLevel }) {
  const strengthItem = getEditOption(EDIT_STRENGTHS, strength)
  const styleItem = getEditOption(EDIT_STYLES, style)
  const levelItem = getEditOption(EDIT_LANGUAGE_LEVELS, languageLevel)
  return {
    strength: strengthItem,
    style: styleItem,
    languageLevel: levelItem,
    summary: `${strengthItem.label} · ${styleItem.label} · ${levelItem.label}`,
  }
}

export const PARAPHRASE_LEVELS = [
  { id: 'simple', label: 'Simple' },
  { id: 'natural', label: 'Natural' },
  { id: 'academic', label: 'Academic' },
  { id: 'toefl', label: 'TOEFL' },
  { id: 'professional', label: 'Professional' },
]

export const WRITING_TOOL_TRACKS = [
  'writing_edit_coach',
  'writing_edit_generate',
  'writing_paraphrase_coach',
  'writing_paraphrase_generate',
  'writing_paraphrase_check',
  'sentence_builder_coach',
  'paragraph_builder_coach',
  'writing_lesson_coach',
  'writing_prompt_outline_coach',
]

export function buildEditMessage({ text, strength, style }) {
  return `[Writing edit request]

Original text:
${text.trim()}

Edit strength: ${strength}
Target style: ${style}

Edit supportively. Preserve the learner's ideas. Return the required Markdown sections.`
}

export function buildParaphraseMessage({ text, level }) {
  return `[Paraphrase request]

Original text:
${text.trim()}

Target level: ${level}

Teach the learner — explain vocabulary and structure changes. Return the required Markdown sections.`
}

export function buildSentenceBuilderMessage({ sentence }) {
  return `[Sentence builder request]

Basic sentence:
${sentence.trim()}

Build corrected, expanded, and stronger versions. Teach the reusable pattern. Return the required Markdown sections.`
}

export function buildParagraphBuilderMessage({ topic, reason, example, explanation, conclusion }) {
  return `[Paragraph builder request]

Topic sentence / main idea:
${topic.trim()}

Reason:
${reason.trim()}

Example:
${example.trim()}

Explanation:
${explanation.trim()}

Conclusion:
${conclusion.trim()}

Use the learner's ideas. Return the required Markdown sections.`
}

export function buildLessonPracticeMessage({ lesson, attempt, level = 'B1' }) {
  return `[Writing lesson mini practice]

Lesson: ${lesson.title}
Skill goal: ${lesson.skillGoal}
Level: ${level}
Practice task: ${lesson.miniPractice.prompt}
${lesson.miniPractice.starter ? `Starter: ${lesson.miniPractice.starter}` : ''}

Student attempt:
${attempt.trim()}

Evaluate supportively. Keep feedback short. Return Corrected version, Why, and Pattern sections.`
}

export function buildPromptOutlineMessage(task) {
  return `[Writing prompt outline request]

Mode: ${task.mode}
Level: ${task.level}
Prompt: ${task.prompt}
Target word count: ${task.wordMin}–${task.wordMax}
Writing goal: ${task.goal}

Generate a guided step-by-step outline, sentence starters, sample opening, and draft paragraph.
Use realistic learner-level English. Return the required Markdown sections.`
}

File: tutor/prompts/writing_tools.py
"""Prompt templates for interactive writing tools (Phase 1)."""

from tutor.prompts.writing_edit import WRITING_EDIT_SYSTEM_PROMPT, WRITING_EDIT_GENERATE_SYSTEM

WRITING_EDIT_COACH_PROMPT = WRITING_EDIT_SYSTEM_PROMPT
WRITING_EDIT_GENERATE_PROMPT = WRITING_EDIT_GENERATE_SYSTEM

WRITING_PARAPHRASE_COACH_PROMPT = """You are an American English teacher teaching paraphrasing to a Persian-speaking learner.

The student message includes original text and a target level (simple, natural, academic, TOEFL, or professional).

RULES:
- Teach the learner — do not only rewrite.
- Explain vocabulary and structure changes clearly.
- Use simple English for simple level; optional short Persian notes when helpful.
- Return clean Markdown only.

Return these sections:

## Simple paraphrase

## Natural paraphrase

## Academic paraphrase

## What changed and why

## Useful phrases

## Now you try

Give one short practice sentence for the learner to paraphrase on their own."""

SENTENCE_BUILDER_COACH_PROMPT = """You are an American English teacher helping a learner build stronger sentences.

The student provides a basic sentence. Improve it step by step while teaching the pattern.

RULES:
- Be supportive and clear.
- American English focus.
- Return clean Markdown only.

Return these sections:

## Basic sentence

## Corrected sentence

## Expanded sentence

## Stronger sentence

## Explanation

## Pattern to reuse

For important grammar mistakes in the basic sentence, append up to 3 correction blocks:

---CORRECTION---
{"original": "...", "corrected": "...", "reason": "...", "persian": "..."}
---END_CORRECTION---"""

PARAGRAPH_BUILDER_COACH_PROMPT = """You are an American English teacher helping a learner build a coherent paragraph from their ideas.

The student provides:
- Topic sentence / main idea
- Reason
- Example
- Explanation
- Conclusion

RULES:
- Use the learner's ideas — do not replace them with unrelated content.
- Teach paragraph structure with clear connectors.
- American English, suitable for TOEFL/academic learners.
- Return clean Markdown only.

Return these sections:

## Generated paragraph

## Better version

## Structure explanation

## Useful connectors

## Next practice suggestion"""

def _templates(task_type: str, system_prompt: str) -> list[dict]:
    return [
        {
            "title": f"{task_type.replace('_', ' ').title()} (Ollama)",
            "task_type": task_type,
            "provider": "ollama",
            "model_name": "qwen2.5:7b",
            "system_prompt": system_prompt,
            "temperature": 0.5,
            "max_tokens": 2000,
            "is_active": True,
        },
        {
            "title": f"{task_type.replace('_', ' ').title()} (OpenAI)",
            "task_type": task_type,
            "provider": "openai",
            "model_name": "gpt-4o-mini",
            "system_prompt": system_prompt,
            "temperature": 0.5,
            "max_tokens": 2000,
            "is_active": True,
        },
        {
            "title": f"{task_type.replace('_', ' ').title()} (Anthropic)",
            "task_type": task_type,
            "provider": "anthropic",
            "model_name": "claude-3-5-haiku-latest",
            "system_prompt": system_prompt,
            "temperature": 0.5,
            "max_tokens": 2000,
            "is_active": True,
        },
    ]


WRITING_TOOLS_TEMPLATES = [
    *_templates("writing_edit_coach", WRITING_EDIT_COACH_PROMPT),
    *_templates("writing_edit_generate", WRITING_EDIT_GENERATE_PROMPT),
    *_templates("writing_paraphrase_coach", WRITING_PARAPHRASE_COACH_PROMPT),
    *_templates("sentence_builder_coach", SENTENCE_BUILDER_COACH_PROMPT),
    *_templates("paragraph_builder_coach", PARAGRAPH_BUILDER_COACH_PROMPT),
]
